ENH: option to export df to Stata dataset with value labels

lmcindewar opened the pull request:
GH38454

- [x] closes #38454
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

lmcindewar committed:
ENH: option to export df to Stata dataset with value labels

GH38454

lmcindewar committed:
Removing unnecessary list comprehension, flake8

lmcindewar committed:
Adding value_labels argument to DataFrame to_stata method

lmcindewar commented:
I added a `StataNonCatValueLabel` to format value labels for non categorical variables. In the `StataWriter`:
* added `value_labels` as optional argument. The format is as close to what is returned from `StataReader.value_labels()`, but the keys should be column names instead of column labels
* renamed `_is_col_cat` to `_has_value_labels`
* `_prepare_non_cat_value_labels` formats the value labels and checks that columns exist and are numeric

bashtage commented on pandas/core/frame.py:
```diff
@@ -2301,6 +2301,7 @@ def to_stata(
         time_stamp: datetime.datetime | None = None,
         data_label: str | None = None,
         variable_labels: dict[Hashable, str] | None = None,
+        value_labels: dict[str, dict[float | int, str]] | None = None,
```
Since these are not keyword only, I don't think this can be inserted here since it could produce a crash ins otherwise correct code. Would be great to move most of these to be keyword only, but that is a different issue.

bashtage replied:
I still think this might need to move to the end of the list. Ideally you would make it keyword only, e.g.,

```
,
*,
value_labels: dict[Hashable, dict[float | int, str]] | None = None,
```

Also note that it should be `dict[Hashable, ...` not `str`

bashtage commented on pandas/core/frame.py:
```diff
@@ -2343,6 +2344,10 @@ def to_stata(
         variable_labels : dict
             Dictionary containing columns as keys and variable labels as
             values. Each label must be 80 characters or smaller.
+        value_labels : dict
```
Should we be more specific here, `dict[hashable, dict]` or `dict of dicts`

bashtage commented on pandas/io/stata.py:
```diff
@@ -3454,6 +3569,7 @@ def __init__(
         time_stamp: datetime.datetime | None = None,
         data_label: str | None = None,
         variable_labels: dict[Hashable, str] | None = None,
+        value_labels: dict[str, dict[float | int, str]] | None = None,
```
Isn't this `dict[Hashable, ...]` since column names don't have to be strings (IIRC non strings are converted/munged until they satisfy Stata reqs)

bashtage commented on pandas/io/stata.py:
```diff
@@ -2222,17 +2295,46 @@ def _write_bytes(self, value: bytes) -> None:
+        for labname, labels in self._non_cat_value_labels.items():
+            if labname not in data.columns:
+                # Value label should apply to a column
+                raise ValueError(
```
KeyError?

bashtage commented on pandas/io/stata.py:
```diff
@@ -2222,17 +2295,46 @@ def _write_bytes(self, value: bytes) -> None:
+                    "can only be applied to numeric columns."
+                )
+            svl = StataNonCatValueLabel(labname, labels)
+            self._value_labels.append(svl)
```
Would a dict be simpler here with key labname?

lmcindewar replied:
Would the dict be for `_value_labels` as { labname: StataValueLabel }? I added the non-categorical value labels onto the existing functions/data used to convert the categoricals columns as much as possible. I'm happy to change this if that makes more sense

bashtage replied:
Probably not worth the effort.

bashtage commented on pandas/io/stata.py:
```diff
@@ -734,6 +735,71 @@ def generate_value_label(self, byteorder: str) -> bytes:
+        self,
+        labname: str,
+        value_labels: dict[float | int, str],
+        encoding: str = "latin-1",
```
Literal["latin-1","utf-8"]? (or maybe some of the other spellings too, e.g., utf8.

bashtage reviewed (changes requested):
Aside from these points, which are mostly minor:

1. Have you created a dta with the full range of supported column types and a mix of fully and partially labeled columns and verified it is correct in Stata? Round tripping is a good start, but it is also necessary to directly verify that the dta renders correctly in Stata.
2. What happens when column names are not legal Stata variable names (e.g., they need to be munged by pre/postpending _). I don't see anything that explicitly hadnles this case, and it might be automagicaly handeled.  It would be good to add a test where a few columns had ad names.

lmcindewar committed:
Updating types and changing ValueError to KeyError for missing column

lmcindewar commented:
Thanks for the feedback! I made the typing changes and will go ahead and create a sample dta for some additional tests

bashtage commented:
You need to import Literal when type checking, and make sure `from __future__ import annotations` is at the top of the file.

An example:

https://github.com/pandas-dev/pandas/blob/accf345fdd4fb92056a5e62d4105fc8f4491ff2c/pandas/core/series.py#L141-L144

lmcindewar committed:
Using converted names for invalid Stata variable names

lmcindewar committed:
Moving value_labels to key word only for to_stata

lmcindewar committed:
Adding tests for invalid Stata names and repeated value labels

lmcindewar committed:
Fixing Literal import

lmcindewar commented:
1. I created a variety of labelled types and checked that labels are being applied and displayed correctly in Stata:

![image](https://user-images.githubusercontent.com/16855568/116783227-9399ad00-aa5b-11eb-9113-a738c3b8d8d6.png)

![image](https://user-images.githubusercontent.com/16855568/116783231-9c8a7e80-aa5b-11eb-8028-14f7723983de.png)

2. The column names were being converted to valid Stata variables *before* the value labels were created. I added a check to use `StataWriter._converted_names` before checking the existing columns. I also moved the creation of `_converted_names` above `_prepare_pandas`, it was being populated in prepare pandas and reset to an empty dict in init

lmcindewar committed:
Moving label encoding to method

bashtage commented on pandas/core/frame.py:
```diff
@@ -2386,6 +2388,11 @@ def to_stata(
+            Dictionary containing columns as keys and dictionaries of column value
+            to labels as values. Labels for a single variable must be 32,000
+            characters or smaller.
+
```
Should have a .. versionadded:: 1.3.0

bashtage commented on pandas/io/stata.py:
```diff
@@ -734,6 +739,71 @@ def generate_value_label(self, byteorder: str) -> bytes:
+    ----------
+    labname : str
+        Value label name
+    value_labels: Dictionary
```
`dict`

bashtage commented on pandas/io/stata.py:
```diff
@@ -734,6 +739,71 @@ def generate_value_label(self, byteorder: str) -> bytes:
+
+        self.labname = labname
+        self._encoding = encoding
+        self.value_labels = list(value_labels.items())
```
Combine these two with `sorted(values_labels.items(),key=lambda ...)`

bashtage replied:
Is this `list[tuple[int | float, str], ...]`?

bashtage commented on pandas/io/stata.py:
```diff
@@ -734,6 +739,71 @@ def generate_value_label(self, byteorder: str) -> bytes:
+        for vl in self.value_labels:
+            category = vl[1]
+            if not isinstance(category, str):
+                category = str(category)
```
Should you convert or raise?  I think raising TypeError feels more explicit since these have to be strings don't they?

bashtage commented on pandas/io/stata.py:
```diff
@@ -734,6 +739,71 @@ def generate_value_label(self, byteorder: str) -> bytes:
+            category = category.encode(encoding)
+            offsets.append(self.text_len)
+            self.text_len += len(category) + 1  # +1 for the padding
+            values.append(vl[0])
```
How do you know this is an int?  If my logic is right, it looks like this is `float | int`, no?

bashtage reviewed (changes requested):
It looks good.  A few final questions. Probably all correct and I'm just not seeing it.

lmcindewar committed:
Updates from review: typing, documentation

lmcindewar commented on pandas/io/stata.py:
```diff
@@ -644,12 +649,22 @@ def __init__(self, catarray: Series, encoding: str = "latin-1"):
 
+        self._prepare_value_labels()
+
+    def _prepare_value_labels(self):
```
I had originally copied the whole init from `StataValueLabel` and only changed where the label name and values were coming from. I moved the encoding of the labels to a separate function here to avoid repeating that process and to make sure both are encoded in the same way
Raising the type error makes sense to me, but it would change the existing categorical encoding. I can remove the addition of `_prepare_value_labels` and only raise the error in the new StataNonCatValueLabel, or make the change in both?

lmcindewar replied:
Hey @bashtage, I realized that I didn't respond directly to your comments. Does this make sense?

bashtage replied:
Yes, it seems clear now. Thanks.

lmcindewar committed:
Fixing mypy errors

bashtage commented on pandas/io/stata.py:
```diff
@@ -3402,6 +3512,11 @@ class StataWriterUTF8(StataWriter117):
 
+    value_labels : dict of dicts
+        Dictionary containing columns as keys and dictionaries of column value
+        to labels as values. Labels for a single variable must be 32,000
```
Is this the total length of all labels for a single variable mus tbe 32000 or smaller, or each label must be 32k or smaller?  Small clarification would be useful.

bashtage commented on pandas/io/stata.py:
```diff
@@ -2222,17 +2286,52 @@ def _write_bytes(self, value: bytes) -> None:
 
+    def _prepare_non_cat_value_labels(self, data: DataFrame) -> None:
+        """
+        Check for value labels provided for non-categorical columns. Value
```
Somethign fishy in this docstring. Possible 2 extra words?

bashtage commented on pandas/io/stata.py:
```diff
@@ -2138,6 +2193,11 @@ class StataWriter(StataParser):
 
+    value_labels : dict of dicts
+        Dictionary containing columns as keys and dictionaries of column value
+        to labels as values. Labels for a single variable must be 32,000
```
Same clarification here.

bashtage commented on pandas/io/stata.py:
```diff
@@ -65,6 +67,9 @@
 
 from pandas.io.common import get_handle
 
+if TYPE_CHECKING:
```
Do you need a check that the Python is 3.8 or later here, with an import from typing_extensions if 3.7?  You can do it like it is done everywhere else, so if this is standard then this is ok (means can't type check on 3.7 unless changed). For an example that supports 3.7,

https://github.com/numpy/numpy/blob/bfb1ac86588456a78af63fa314b768f4ebc764ff/numpy/random/_generator.pyi#L47

jreback replied:
you can just import this from pandas._typing (so when we change to 3.8 this would be fine)

bashtage reviewed (changes requested):
Really minor clarifications.  Ping when done and I'll allow CI to run, and then we can see about merging.

lmcindewar committed:
Clarifying comment on label length

bashtage commented:
If the CI is green, then we'll get it merged.  Thanks.

lmcindewar commented:
> If the CI is green, then we'll get it merged. Thanks.

@bashtage I can't tell where things are failing. Any suggestions?

jreback commented on pandas/io/stata.py:
```diff
@@ -644,24 +649,34 @@ def __init__(self, catarray: Series, encoding: str = "latin-1"):
         self.text_len = 0
         self.txt: list[bytes] = []
         self.n = 0
+        self.off = np.array([])
```
can you add a comment on this (what ar ethese) are these integer types?

jreback commented on pandas/io/stata.py:
```diff
@@ -734,6 +749,46 @@ def generate_value_label(self, byteorder: str) -> bytes:
+    labname : str
+        Value label name
+    value_labels: Dictionary
+        Mapping of values to labels
```
might be easier to add a method in the super class to avoid overly defining this (e.g. you are repeating a bunch). is the signature the same? then you don't need an extended doc-string

lmcindewar replied:
I wasn't sure how to best deal with the duplication, the only difference between the existing StataValueLabel and new non-categorical value label is the init: the non-categorical takes the label name and a dict of value labels directly, the existing `StataValueLabel` is created from a categorical series and pulls the name and labels from it.

The two options I've thought of for getting rid of the duplication are:
* change `StataValueLabel` to init from a label name and value_labels dictionary then add a `from_cat` method to extract the name and label from a Categorical
* add labname and value_labels to the StataValueLabel init and let it be created from either those or the catarray

Both would get rid of the new non categorical label class entirely, the first would be a breaking change to StataValueLabel. Would one of these be acceptable or is there a better way that I'm not thinking of?

lmcindewar replied:
@jreback I moved the duplicated attribute definitions to _prepare_value_labels. Would either of the changes from above be better?

jreback commented on pandas/io/stata.py:
```diff
@@ -2188,15 +2248,20 @@ def __init__(
         self._data_label = data_label
         self._variable_labels = variable_labels
+        self._non_cat_value_labels = value_labels
+        self._value_labels: list[StataValueLabel] = []
```
why don't you rename this to _cat_value_labels to distingish and make consistent

lmcindewar replied:
_value_labels is a combined list of categorical and non-categorical labels. there isn't a standalone _cat_value_labels because they're added straight to _value_labels after converting categoricals in the dataframe

jreback commented on pandas/io/stata.py:
```diff
@@ -2222,17 +2286,52 @@ def _write_bytes(self, value: bytes) -> None:
         """
         self.handles.handle.write(value)  # type: ignore[arg-type]
 
+    def _prepare_non_cat_value_labels(self, data: DataFrame) -> None:
```
this is modifying global state and is pretty opaque to the reader. ideally these would return values and state would be changed in the caller.

lmcindewar replied:
all of the state change to the caller.

bashtage commented:
Close and reopen to triger CI.

bashtage commented:
@lmcindewar Can you take a look @jreback 's comments?

bashtage commented:
Failure is a random one due to codecov.

lmcindewar committed:
Removing duplication in value label class and returning labels from prepare_non_cat

bashtage commented:
Typo in docstring

```
-        """ Encode value labels. """
+        """Encode value labels."""
```

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -2386,6 +2388,13 @@ def to_stata(
+            to labels as values. Labels for a single variable must be 32,000
+            characters or smaller.
+
+            .. versionadded:: 1.3.0
```
probably 1.4

simonjayhawkins commented on pandas/io/stata.py:
```diff
@@ -2138,6 +2188,11 @@ class StataWriter(StataParser):
+        Dictionary containing columns as keys and dictionaries of column value
+        to labels as values. The combined length of all labels for a single
+        variable must be 32,000 characters or smaller.
+
```
versionadded tag here too

simonjayhawkins reviewed (changes requested):
can you also add a release note (1.4 I think)

lmcindewar committed:
Adding versionaddeds

lmcindewar committed:
Typo in spacing of docstring

lmcindewar committed:
Adding release note

lmcindewar commented:
@simonjayhawkins I'd completely missed your comment, I added a short release note. I think that should be it unless there's anything left from the previous reviews

simonjayhawkins commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -34,6 +34,7 @@ Other enhancements
 -  Additional options added to :meth:`.Styler.bar` to control alignment and display, with keyword only arguments (:issue:`26070`, :issue:`36419`)
 - :meth:`Styler.bar` now validates the input argument ``width`` and ``height`` (:issue:`42511`)
 - :meth:`Series.ewm`, :meth:`DataFrame.ewm`, now support a ``method`` argument with a ``'table'`` option that performs the windowing operation over an entire :class:`DataFrame`. See :ref:`Window Overview <window.overview>` for performance and functional benefits (:issue:`42273`)
+- :meth:`DataFrame.to_stata` and :meth:`StataWriter` now accept the keyword only argument ``value_labels`` to save labels for non-categorical columns
```
```suggestion
- :meth:`DataFrame.to_stata` and :meth:`StataWriter` now accept the keyword only argument ``value_labels`` to save labels for non-categorical columns (:issue:`38454`)
```

bashtage commented:
LGTM.

bashtage reviewed (approved):
All issues have been addressed and once green looks good to merge.

lmcindewar commented:
@bashtage I haven't been able to nail down what's causing these type errors. It passes type checking with an added annotation on line 2542 in io/stata.py:
```python
self.data: DataFrame = data
```
but I wasn't sure if there was an underlying issue that needs to be fixed

bashtage commented:
Try adding `self.data = data` after the super in `__init__`

lmcindewar committed:
WSetting data in Statawriter init

bashtage commented:
Failures seem unrelated to anything in this PR.

bashtage commented:
Can you rebase on master to fix the merge conflict. and then I think this is ready.

bashtage commented:
I rebased it.

bashtage commented:
@jreback I think this PR is in good shape and should be merged.  3 unrelated failures in doc tests.

bashtage commented:
@lmcindewar can you rebase so we can look into merging?

lmcindewar commented:
@bashtage I rebased it, let me know if there's anything else needed to merge it

jreback committed:
Merge branch 'master' into GH38454-export-stata-value-labels

jreback reviewed (approved):
lgtm. over to you @bashtage 

@lmcindewar would be nice to add an example to the doc-string as well (can be here or in a followup)

bashtage commented:
Thanks @lmcindewar .  It would be great if you could do an example in a follow-up.

lmcindewar commented:
@bashtage Will do. Thanks again for all of the reviews
